R/plot_variable_importance.R

```diff
@@ -0,0 +1,182 @@
+#' or a \code{patchwork} object containing two plots (original and new features).
+#' This is useful for diagnosing which group of features contributes more to the stacked model.
+#'
+#' @param stack_model A \code{caretStack} model trained using \code{caretEnsemble}.
+#' It should have an attribute \code{original_features} listing the original input variables.
+#' If this attribute is missing, all variables are treated as "new".
+#'
+#' @param newdata A data frame containing the data used for calculating variable importance.
+#' Typically this should be the validation or test set.
+#'
+#' @param stat_type Optional character string indicating which summary statistic
+#' of the opposite group to include as a gray bar for reference.
+#' Must be one of \code{"mean"}, \code{"sum"}, or \code{"max"}. If \code{NULL}, no statistic is shown.
+#' If invalid, an error is thrown.
+#'
+#' @return A \code{ggplot} object if no original features attribute is found,
+#' or a \code{patchwork} object with two bar plots:
+#' one for original features (in blue) and one for new features (in red).
+#' If \code{stat_type} is provided, a gray bar appears in each plot representing
+#' the selected summary statistic from the opposite group (e.g., mean of new features
+#' shown in the original features plot).
+#'
+#' @details
+#' - Variable importance is computed using \code{caret::varImp}.
+#' - If the model lacks the \code{original_features} attribute, all variables are considered new.
+#' - Requires the packages: \code{data.table}, \code{ggplot2}, \code{caret}, and \code{patchwork}.
+#'
+#' @importFrom data.table :=
+#' @export
+plot_variable_importance <- function(stack_model, newdata, stat_type = NULL) {
+  # Prepare importance table
+  imp_dt <- prepare_importance(stack_model, newdata)
+  original_features <- stack_model[["original_features"]]
+
+  # Validate stat_type
+  valid_stats <- c("mean", "sum", "max")
+  if (!is.null(stat_type) && !(stat_type %in% valid_stats)) {
+    stop("stat_type must be 'mean', 'sum', or 'max'.", call. = FALSE)
+  }
+
+  # Case 1: No original features → all variables treated as "New"
+  if (is.null(original_features)) {
+    imp_dt[, type := "New"]
+    imp_dt[, is_stat := FALSE]
+    plot_group(
+      imp_dt, "Variable Importance - New Features", c("FALSE" = "red", "TRUE" = "gray"),
+      max(imp_dt$weight, na.rm = TRUE)
+    )
+  } else {
+    # Case 2: Separate original and new, optionally add cross-group statistics
+    imp_list <- add_cross_group_stats(imp_dt, original_features, stat_type)
+    max_weight <- max(c(imp_list$imp_original$weight, imp_list$imp_new$weight), na.rm = TRUE)
+
+    p1 <- plot_group(imp_list$imp_original, "Original Features", c("FALSE" = "blue", "TRUE" = "gray"), max_weight)
+    p2 <- plot_group(imp_list$imp_new, "New Features", c("FALSE" = "red", "TRUE" = "gray"), max_weight)
+
+    # Combine plots side by side
+    p1 + p2 + patchwork::plot_layout(ncol = 2L)
```

Comment:
Can you use [facet_grid](https://ggplot2.tidyverse.org/reference/facet_grid.html) to avoid the patchwork dep?